### PR TITLE
[tiff] fix invalid memory accesses

### LINKED ISSUE #24120: [tiff] fix invalid memory accesses
****

**Is your feature request related to a problem? Please describe.**
A number of invalid memory accesses were reported to NIST

**Proposed solution**
Apply patches to resolve these invalid memory accesses that were merged into libtiff's master branch

**Describe alternatives you've considered**
Ignore the issue. Most of the issues affect only the **tiffcrop** application which we don't even install unless the **tools** option is enabled

**Additional context**
https://nvd.nist.gov/vuln/detail/CVE-2022-0561
https://nvd.nist.gov/vuln/detail/CVE-2022-0562
https://nvd.nist.gov/vuln/detail/CVE-2022-0865
https://nvd.nist.gov/vuln/detail/CVE-2022-0891
https://nvd.nist.gov/vuln/detail/CVE-2022-0907
https://nvd.nist.gov/vuln/detail/CVE-2022-0908
https://nvd.nist.gov/vuln/detail/CVE-2022-0909
https://nvd.nist.gov/vuln/detail/CVE-2022-0924
https://nvd.nist.gov/vuln/detail/CVE-2022-22844

### PR DESCRIPTION
**Pull request by SchaichAlonso**
**Describe the pull request**

- #### What does your PR fix?
  Fixes #24120 

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  triplets unchanged, CI baseline not modified

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/tiff/portfile.cmake [hunk 1]
```diff
@@ -10,6 +10,15 @@ vcpkg_from_gitlab(
...
     FindCMath.patch
+    fix-CVE-2022-0561.diff
+    fix-CVE-2022-0562.diff
+    fix-CVE-2022-0865.diff
+    fix-CVE-2022-0891.diff
+    fix-CVE-2022-0907.diff
+    fix-CVE-2022-0908.diff
+    fix-CVE-2022-0909.diff
+    fix-CVE-2022-0924.diff
+    fix-CVE-2022-22844.diff
```
[the remark this answers is not in the document]
Comment (mathisloge) [reply ports/tiff/portfile.cmake:21]: those are linked in the CVE-* links

[the remark this answers is not in the document]
Comment (JonLiu1993) [reply ports/tiff/portfile.cmake:21]: @SchaichAlonso ,Thanks for your pr, are these changes allowed by upstream? There are too many changes involving source, we recommend to notify upstream first

**Comment by JonLiu1993** [comment]
@SchaichAlonso ,Thanks for your pr, are these changes allowed by upstream? There are too many changes involving source, we recommend to notify upstream first

File: ports/tiff/portfile.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/tiff/portfile.cmake:21]: Yeah but I have to do:

Google the CVE -> Go to CVE site -> find link to commit -> find MR for commit

Having the MR numbers would be

go to tiff gitlab -> directly insert MR nummer to go to MR

**Comment by Neumann-A** [comment]
> There are too many changes involving source, we recommend to notify upstream first

Probably better to download the upstream diffs instead of adding it to vcpkg.

File: ports/tiff/portfile.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathisloge) [reply ports/tiff/portfile.cmake:21]: he has linked the CVE`s in the commit message so it's `Go to CVE site -> find link to commit -> find MR for commit` :D 

jokes aside. `go to tiff gitlab -> directly insert MR nummer to go to MR` would be easier.

[the remark this answers is not in the document]
Comment (SchaichAlonso) [reply ports/tiff/portfile.cmake:21]: Hi

Commits merged to libtiff master via Merge Requests are
CVE-2022-0865 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/306
CVE-2022-0891 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/307
CVE-2022-0907 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/314
CVE-2022-0909 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/310
CVE-2022-0924 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/311
CVE-2022-22488 -> https://gitlab.com/libtiff/libtiff/-/merge_requests/287

Commits merged to libtiff master w/o merge request:
CVE-2022-0561 -> https://gitlab.com/libtiff/libtiff/-/commit/eecb0712f4c3a5b449f70c57988260a667ddbdef
CVE-2022-0562 -> https://gitlab.com/libtiff/libtiff/-/commit/561599c99f987dc32ae110370cfdd7df7975586b
CVE-2022-0908 -> https://gitlab.com/libtiff/libtiff/-/commit/a95b799f65064e4ba2e2dfc206808f86faf93e85

[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/tiff/portfile.cmake:21]: > he has linked the CVE

I think that is a github feature

**Comment by SchaichAlonso** [comment]
@JonLiu1993 The patches I've added have all been already accepted by the upstream, and are in fact copies of commits made into upstream's master branch.

@Neumann-A I agree that it would probably be better to reference a list of commits to cherry-pick from upstream over adding copies of the commit-diffs here and referencing the copies.

However ... I have no clue how to do that, and the documentation of `vcpkg_from_gitlab` doesn't explain how to do it, either.

**Comment by Neumann-A** [comment]
@SchaichAlonso 

Other ports do:

```
vcpkg_download_distfile(patch1679
    URLS "https://patch-diff.githubusercontent.com/raw/apache/incubator-brpc/pull/1679.diff"
    FILENAME "apache-incubator-brpc-1679.diff"
    SHA512 4b1e5717b44aa6a741ddd49b1408e3e556f6d845d5e8a5cfccf2f2d7ebe39aed19c3dad703db7a9ebd0446ac1f225e7dbdd2ff1f23f34fd60c3ef59aaa07b789
)
```

**Review by JonLiu1993 — APPROVED** [review]

**Comment by dg0yt** [comment]
> ~~~
> vcpkg_download_distfile(patch1679
>     URLS "https://patch-diff.githubusercontent.com/raw/apache/incubator-brpc/pull/1679.diff"
>     FILENAME "apache-incubator-brpc-1679.diff"
>     SHA512 4b1e5717b44aa6a741ddd49b1408e3e556f6d845d5e8a5cfccf2f2d7ebe39aed19c3dad703db7a9ebd0446ac1f225e7dbdd2ff1f23f34fd60c3ef59aaa07b789
> )

IIRC this approach broke not so long ago: GH may change the patch format, and then the SHA512 no longer matches. (But GH doesn't even promise of stability tarballs for arbitrary references. Only release tarballs are promised to be stable.)

**Comment by JonLiu1993** [comment]
@dg0yt ,Do you have any good advice?

**Comment by dg0yt** [comment]
Obviously my advice is now just the opposite of what was said and done before. It is up to the "maintainers" to decide.

I have to admit that these patches come from gitlab, not github. Maybe it is more stable, but I don't know.

The best way out would be a new upstream patch release, but they don't want to be pushed from consumers AFAIU https://www.awaresystems.be/imaging/tiff/tml/tm202203.html.
FTR, at least one change seem to include a feature improvement (e.g. https://gitlab.com/libtiff/libtiff/-/merge_requests/311.diff). 

How many other security fixes are lacking from `HEAD`?

**Comment by JonLiu1993** [comment]
@strega-nil-ms, Could you please take a look?

**Review by JonLiu1993 — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
We need to talk about what we want to do in these kinds of situations; we'll discuss it today.

**Comment by strega-nil-ms** [comment]
vcpkg team agrees this is good, thanks!

**Pull request merged**